### DISCUSSION TITLE
Accessing a NATS Image inside multiple image Kubernetes deployment from the deployment

### DISCUSSION THREAD

**Discussion by oheckmann** [nats-io/nats.go#1210] [Q&A]
I am trying to consolidate several services which were previously run as different pods in Kuberentes, with the hope of reducing system latency. Our system involves communication between several different pods which were often on different nodes in our cluster - due to the specific element of the deployment (most notably the large size of the data sent between the programs), communication between programs on different nodes is the primary source of latency in our system.

One of the components is NATS, which handles the communication between the different services. I naively tried to include the NATS image in the deployment like so:

    ---
    apiVersion: apps/v1
    kind: Deployment
    metadata:
      name: vei-deployment
    spec:
      replicas: 1
      selector:
        matchLabels:
          app: camera-system
      template:
        metadata:
          labels:
            app: camera-system
        spec:
          containers:
            - name: application
              image: image1:latest
            - name: application2
              image: image2:latest
            - name: application3
              image: image3:latest
            - name: nats
              image: nats

Application 1 is a Golang program that was connects to NATs using the standard library, and the code:

    //start a new connection with NATS
    var nc, err = nats.Connect("nats://nats:4222")
    if err != nil {
            panic(err)
    }
    defer nc.Close()

However, when doing so, the systems refuse to connect to NATs, throwing the error `panic: dial tcp: lookup nats on 10.43.0.10:53: no such host
`, though according to the logs, NATs is working correctly. What am I missing in order to make this deployment operable?

**Comment by wallyqs** [comment]
In your example above you are running the NATS container within the same Pod, if that is intentional then in order to connect to it you would have to connect to it like `nats.Connect("localhost")` since not backed by a service. Usually NATS is ran as a headless Service (which would result in the addressable `"nats://nats:4222"`), here you can find the Helm chart and how to deploy it as a service: https://github.com/nats-io/k8s/tree/main/helm/charts/nats

**Reply by oheckmann** [reply]
Yes, I am aware that what I am doing is not good practice in Kubernetes, I am just in a situation in which the specific benefits may outweigh the costs. 
As for the container, I don't need to use the specific port number? In this case, how would I use something like port 8222 in the headless service?

**Reply by wallyqs** [reply]
the port 8222 is for monitoring so not needed for the nats traffic.  by default `nats.Connect("localhost")` becomes expanded in the client to the default port like `nats.Connect("nats://localhost:4222")`
